### PR TITLE
Use `moduleResolution: 'node16'`

### PR DESCRIPTION
**Pull request by Princesseuh**
## Changes

**Blocked by https://github.com/vitejs/vite/pull/13947**

https://github.com/withastro/astro/pull/7780 made it so we use `moduleResolution: 'bundler'`, this is mostly fine but since we still require extensions in the runtime, it's better for us to use `moduleResokution: 'node16'` instead. However, this has, a lot of requirements. This PR attempts to do it

## Testing

Tests should pass!

## Docs

N/A

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+  "typescript.preferences.importModuleSpecifierEnding": "js"
```
Comment (bluwy) [inline .vscode/settings.json:2]: I just realized I set this per your recommendation but with `javascript.` 🤦 That's why it was not working.

**Review by bluwy — COMMENTED** [review]
Besides the Vite issue, the changes looks good to me. Thanks for tackling this!

**Comment by bluwy** [comment]
Not sure if Vite is the only blocker for node16, but it has been fixed and released in 4.4.7

**Comment by Princesseuh** [comment]
Closing this, will have to restart from scratch pretty much

**Pull request closed without merging**